File: gene2protein_akt1.py
import pyensembl
import requests
import json
import os
import pandas as pd
from typing import List, Dict, Any

# ==============================================================================
# Assumptions & Context
# ==============================================================================
# 1. Goal: The assignment asks to identify transcripts and proteins for a specific
#    gene (AKT1, ENSG00000142208) and map the protein coordinates back to the
#    genome.
# 2. Tool Translation: The professor suggested using the R package `ensembldb`.
#    Since you requested a Python solution, this script uses `pyensembl` for local
#    gene/transcript queries and the Ensembl REST API for the more complex
#    protein-to-genome coordinate mapping, which `pyensembl` doesn't handle natively.
# 3. Species & Assembly: Assumed Human (Homo sapiens), Assembly GRCh38 (Release 109),
#    as this is standard unless otherwise specified.
# 4. Input Gene: AKT1 (ENSG00000142208).
# 5. Output: Tab-separated files (TSV) listing transcripts, proteins, and the
#    genomic coordinates for the protein sequences.
# ==============================================================================

# Configuration
GENE_ID = "ENSG00000142208"  # AKT1
GENE_NAME = "AKT1"
ASSEMBLY_RELEASE = 109
SPECIES = "homo_sapiens"

# File output paths
TRANSCRIPTS_OUT = f"{GENE_NAME}_transcripts.tsv"
PROTEINS_OUT = f"{GENE_NAME}_proteins.tsv"
MAPPING_OUT = f"{GENE_NAME}_protein_to_genome_map.tsv"

def setup_pyensembl() -> pyensembl.EnsemblRelease:
    """
    Initializes pyensembl and downloads the necessary reference data if not present.
    """
    print(f"Setting up Ensembl Release {ASSEMBLY_RELEASE}...")
    ensembl = pyensembl.EnsemblRelease(ASSEMBLY_RELEASE)
    
    # Check if we need to download/install data
    try:
        ensembl.gene_by_id(GENE_ID)
    except ValueError:
        print("Reference data not found. Downloading and installing (this may take a while)...")
        ensembl.download()
        ensembl.index()
        print("Installation complete.")
        
    return ensembl

def get_transcripts_and_proteins(ensembl: pyensembl.EnsemblRelease, gene_id: str) -> tuple[List[str], List[str]]:
    """
    Retrieves all transcripts and associated proteins for the given gene.
    """
    print(f"Fetching transcripts and proteins for gene {gene_id}...")
    gene = ensembl.gene_by_id(gene_id)
    transcripts = gene.transcripts
    
    transcript_ids = []
    protein_ids = []
    
    print("\n--- Transcripts & Proteins Found ---")
    for transcript in transcripts:
        t_id = transcript.transcript_id
        p_id = transcript.protein_id
        
        transcript_ids.append(t_id)
        if p_id: # Not all transcripts have a translated protein product
            protein_ids.append(p_id)
            print(f"Transcript: {t_id} -> Protein: {p_id}")
        else:
            print(f"Transcript: {t_id} -> No protein product (non-coding)")
            
    return transcript_ids, protein_ids

def get_protein_to_genome_mapping(protein_id: str) -> List[Dict[str, Any]]:
    """
    Uses the Ensembl REST API to get genomic coordinates for a given protein.
    """
    # Endpoint to map translation (protein) to genome
    server = "https://rest.ensembl.org"
    ext = f"/map/translation/{protein_id}/1..1000000?" # Using a large range to cover the whole protein
    
    url = f"{server}{ext}"
    headers = {"Content-Type": "application/json"}
    
    try:
        response = requests.get(url, headers=headers)
        if not response.ok:
            if response.status_code == 400 and "not a valid" in response.text:
                 # Sometimes obsolete IDs linger, or API limits
                 print(f"  Warning: Protein ID {protein_id} not found via REST API.")
                 return []
            response.raise_for_status()
            
        data = response.json()
        mappings = []
        
        # The API returns mappings of the queried sequence (protein coordinates)
        # to the target sequence (genomic coordinates).
        if 'mappings' in data:
             for m in data['mappings']:
                 # We only care about mapped regions (not gaps/unmapped)
                 if 'mapped' in m:
                    mapped_info = m['mapped']
                    mappings.append({
                        "Protein_ID": protein_id,
                        "Chromosome": mapped_info['seq_region_name'],
                        "Genomic_Start": mapped_info['start'],
                        "Genomic_End": mapped_info['end'],
                        "Strand": mapped_info['strand'],
                        "Protein_Start": m['original']['start'], # AA coordinate start
                        "Protein_End": m['original']['end']      # AA coordinate end
                    })
        return mappings
        
    except requests.exceptions.RequestException as e:
        print(f"  Error fetching mapping for {protein_id}: {e}")
        return []


def main():
    print(f"Starting analysis for {GENE_NAME} ({GENE_ID})")
    
    # 1. Setup PyEnsembl
    ensembl = setup_pyensembl()
    
    # 2. Get Transcripts and Proteins
    transcript_ids, protein_ids = get_transcripts_and_proteins(ensembl, GENE_ID)
    
    # Save Transcripts
    pd.Series(transcript_ids, name="Transcript_ID").to_csv(TRANSCRIPTS_OUT, index=False, header=True)
    print(f"\nSaved {len(transcript_ids)} transcripts to {TRANSCRIPTS_OUT}")
    
    # Save Proteins
    pd.Series(protein_ids, name="Protein_ID").to_csv(PROTEINS_OUT, index=False, header=True)
    print(f"Saved {len(protein_ids)} proteins to {PROTEINS_OUT}")
    
    # 3. Generate Protein to Genome Mapping
    print("\nMapping proteins to genome coordinates using Ensembl REST API...")
    all_mappings = []
    
    for p_id in protein_ids:
        print(f"  Processing {p_id}...")
        mapping_data = get_protein_to_genome_mapping(p_id)
        if mapping_data:
            all_mappings.extend(mapping_data)
            
    if all_mappings:
        df_mappings = pd.DataFrame(all_mappings)
        # Sort for neatness
        df_mappings = df_mappings.sort_values(by=["Protein_ID", "Protein_Start"])
        df_mappings.to_csv(MAPPING_OUT, sep='\t', index=False)
        print(f"\nSaved protein-to-genome mappings to {MAPPING_OUT}")
    else:
        print("\nNo mappings were generated. Please check the API responses.")

    print("\nAnalysis Complete.")
    print("\nNext Step for IGV:")
    print(f"1. Open IGV (installed locally).")
    print(f"2. Select the Human hg38 (GRCh38) genome.")
    print(f"3. Search for {GENE_NAME} in the search bar to view your gene.")

if __name__ == "__main__":
    main()

File: gene2protein_akt1.R
##############################################################################
# Gene2Protein Assignment Solution
# Target Gene: AKT1 (ENSG00000142208)
# Environment: VSCode + Conda R Environment
##############################################################################

############################
# Load Required Packages
############################

suppressPackageStartupMessages(library(AnnotationHub))
suppressPackageStartupMessages(library(ensembldb))
suppressPackageStartupMessages(library(IRanges))

cat("\nInitializing AnnotationHub...\n")

############################
# Connect to AnnotationHub
############################

# Increase timeout for slow internet
options(timeout = 300)
options(download.file.method = "libcurl")

cat("Connecting to AnnotationHub...\n")

ah <- tryCatch({

    # Try online connection first
    AnnotationHub(localHub = FALSE)

}, error = function(e) {

    cat("\n[WARNING] Network connection failed.\n")
    cat("Details:\n")
    cat(e$message, "\n")

    cat("Attempting local cache access...\n")

    # Fallback to local cache
    AnnotationHub(localHub = TRUE)
})

cat("Connected successfully.\n")

############################
# Load Ensembl Database
############################

cat("\nFetching EnsDb database...\n")

edb <- tryCatch({

    # Professor specified database
    ah[["AH119325"]]

}, error = function(e) {

    cat("\n[WARNING] AH119325 unavailable.\n")
    cat("Searching for latest Homo sapiens EnsDb...\n")

    human_ensdbs <- query(ah, c("EnsDb", "Homo sapiens"))

    if (length(names(human_ensdbs)) == 0) {

        stop(
            "FATAL ERROR: No EnsDb database found.\n",
            "Run script once with active internet."
        )
    }

    # Use latest database
    fallback_id <- tail(names(human_ensdbs), 1)

    cat(sprintf("Using fallback database: %s\n", fallback_id))

    ah[[fallback_id]]
})

############################
# Define Target Gene
############################

target_gene_id   <- "ENSG00000142208"
target_gene_name <- "AKT1"

cat(sprintf(
    "\nStarting analysis for %s (%s)\n",
    target_gene_name,
    target_gene_id
))

##############################################################################
# 1. FETCH ALL TRANSCRIPTS
##############################################################################

cat("\nFetching transcripts...\n")

txs <- transcripts(
    edb,
    filter = ~ gene_id == target_gene_id
)

# Create dataframe
tx_df <- data.frame(
    Transcript_ID = txs$tx_id
)

# Save transcripts
write.table(
    tx_df,
    file = "AKT1_transcripts.tsv",
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
)

cat(sprintf(
    "Saved %d transcripts to AKT1_transcripts.tsv\n",
    nrow(tx_df)
))

##############################################################################
# 2. FETCH ALL PROTEINS
##############################################################################

cat("\nFetching proteins...\n")

prts <- proteins(
    edb,
    filter = ~ gene_id == target_gene_id
)

# Remove NA or empty IDs
protein_ids <- prts$protein_id[
    !is.na(prts$protein_id) &
    prts$protein_id != ""
]

# Keep unique proteins
protein_ids <- unique(protein_ids)

# Create dataframe
prt_df <- data.frame(
    Protein_ID = protein_ids
)

# Save proteins
write.table(
    prt_df,
    file = "AKT1_proteins.tsv",
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
)

cat(sprintf(
    "Saved %d proteins to AKT1_proteins.tsv\n",
    nrow(prt_df)
))

##############################################################################
# 3. PROTEIN TO GENOME MAPPING
##############################################################################

cat("\nMapping proteins to genome coordinates...\n")

mapping_results <- data.frame()

# Process each protein
for (pid in protein_ids) {

    cat(sprintf("Processing protein: %s\n", pid))

    # Fetch protein data
    protein_data <- proteins(
        edb,
        filter = ~ protein_id == pid
    )

    protein_df <- as.data.frame(protein_data)

    # Skip empty result
    if (nrow(protein_df) == 0) {

        cat(" -> No protein data found\n")
        next
    }

    # Check for protein_length column
    if (!"protein_length" %in% colnames(protein_df)) {

        cat(" -> protein_length column missing\n")
        next
    }

    protein_length <- protein_df$protein_length[1]

    # Validate protein length
    if (
        length(protein_length) == 0 ||
        is.na(protein_length) ||
        protein_length <= 0
    ) {

        cat(" -> Invalid protein length\n")
        next
    }

    # Create IRanges object
    ir <- IRanges(
        start = 1,
        end = protein_length
    )

    # Run protein to genome mapping
    mapped <- tryCatch({

        proteinToGenome(
            ir,
            edb,
            protein_id = pid
        )

    }, error = function(e) {

        cat(sprintf(
            " -> Mapping failed: %s\n",
            e$message
        ))

        return(NULL)
    })

    # Skip failed mappings
    if (is.null(mapped) || length(mapped) == 0) {
        next
    }

    # Convert to dataframe
    tmp_df <- as.data.frame(mapped)

    # Skip empty dataframe
    if (nrow(tmp_df) == 0) {
        next
    }

    # Add protein ID
    tmp_df$Protein_ID <- pid

    # Append results
    mapping_results <- rbind(
        mapping_results,
        tmp_df
    )
}

############################
# Save Mapping Results
############################

if (nrow(mapping_results) > 0) {

    clean_mapping <- data.frame(
        Protein_ID    = mapping_results$Protein_ID,
        Chromosome    = mapping_results$seqnames,
        Genomic_Start = mapping_results$start,
        Genomic_End   = mapping_results$end,
        Strand        = mapping_results$strand
    )

    write.table(
        clean_mapping,
        file = "AKT1_protein_to_genome_map.tsv",
        sep = "\t",
        row.names = FALSE,
        quote = FALSE
    )

    cat("\nSaved mapping file: AKT1_protein_to_genome_map.tsv\n")

} else {

    cat("\nWARNING: No mappings generated.\n")
}

##############################################################################
# FINAL MESSAGE
##############################################################################

cat("\n============================================================\n")
cat("Analysis Complete!\n")
cat("Generated Files:\n")
cat("1. AKT1_transcripts.tsv\n")
cat("2. AKT1_proteins.tsv\n")
cat("3. AKT1_protein_to_genome_map.tsv\n")
cat("============================================================\n")

cat("\nNext Steps:\n")
cat("1. Open IGV locally\n")
cat("2. Set genome reference to GRCh38 / hg38\n")
cat("3. Search for AKT1\n")
cat("4. Compare coordinates with generated mapping file\n\n")